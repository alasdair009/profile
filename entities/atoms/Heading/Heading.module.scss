@use "../../../app/styles/colors";
@use "../../../app/styles/dimensions";
@use "../../../app/styles/lineClamp";
@use "../../../app/styles/typography";
@use "../../../app/styles/helpers/rem";
@use "sass:map";

@mixin append-element($element-name) {
  @at-root #{selector-append($element-name, &)} {
    @content;
  }
}

.root {
  border: none !important;
  color: var(--color) !important;
  font-weight: map.get(typography.$fontWeights, bold);
  margin: 0 auto rem.toRem(map.get(dimensions.$sizes, s24)) !important;
  max-width: rem.toRem(dimensions.$globalTextMaxWidth);
  padding-bottom: 0;
  position: relative;
  text-align: var(--align);
  text-wrap: balance;
  width: fit-content;

  &Clamp {
    @include lineClamp.lineClamp;
  }

  &.Shadow {
    text-shadow: 0 0 rem.toRem(map.get(dimensions.$sizes, s4))
      colors.$colorblackEvil;
  }

  &::after {
    background: colors.$colorGreenGrass;
    bottom: 0;
    content: "";
    display: none;
    height: rem.toRem(3);
    left: 0;
    position: absolute;
    transform: rotate(-2deg);
    width: 100%;
  }

  &.h1 {
    max-width: rem.toRem(map.get(dimensions.$sizes, s512));
    padding-bottom: rem.toRem(map.get(dimensions.$sizes, s8));

    &::after {
      display: block;
    }
  }

  &.h2 {
    max-width: rem.toRem(map.get(dimensions.$sizes, s512));

    &::after {
      display: block;
    }
  }

  @each $headerElement, $headerSize in typography.$headerSizes {
    &.#{$headerElement} {
      font-size: rem.toRem($headerSize);
    }
  }
}
