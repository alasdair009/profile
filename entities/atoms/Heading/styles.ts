import styled from "styled-components";
import {CSSProperties} from "react";
import {globalTextMaxWidth, sizes} from "../../design-tokens/dimensions";
import {colors} from "../../design-tokens/colors";
import {rem} from "polished";
import {headingSizes, HeadingTypes} from "../../design-tokens/typography";
import {Property} from "csstype";

export const Root = styled.h1<{$color: Property.Color, $level: HeadingTypes, $align: Property.TextAlign}>`
  color: ${({$color}) => $color};
  font-size: ${({$level}) => headingSizes[$level]};
  margin: 0 auto ${sizes.s24.rem};
  max-width: ${rem(globalTextMaxWidth)};
  padding-bottom: ${({$level}) => $level === "h1" ? sizes.s8.rem : 0};
  position: relative;
  text-align: ${({$align}) => $align};
  width: fit-content;
  
  &::after {
    background: ${colors.greenGrass};
    bottom: 0;
    content: "";
    display: ${({$level}) => $level === "h1" ? "block" : "none"};
    height: ${rem(3)};
    left: 0;
    position: absolute;
    transform: rotate(-2deg);
    width: 100%;
  }
`;