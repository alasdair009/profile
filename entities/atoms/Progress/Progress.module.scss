@use "../../../app/styles/colors";
@use "../../../app/styles/dimensions";
@use "../../../app/styles/radii";
@use "../../../app/styles/helpers/rem";

.root {
  align-items: center;
  aspect-ratio: 1;
  border-radius: 50%;
  display: flex;
  height: rem.toRem(map-get(dimensions.$sizes, s96));
  justify-content: center;
  position: relative;

  &::before {
    background: radial-gradient(
      transparent,
      transparent 30%,
      rgba(colors.$colorblackEvil, 0.5) 30%,
      rgba(colors.$colorblackEvil, 0.5) 70%,
      transparent 50%
    );
    border-radius: map-get(radii.$borderRadii, round);
    content: "";
    height: 100%;
    left: 0;
    position: absolute;
    top: 0;
    width: 100%;
  }
}

.progressElement {
  appearance: none;
  background: conic-gradient(
    colors.$colorGreenGrass,
    colors.$colorWhiteGhost,
    colors.$colorWhiteGhost calc((var(--value) / var(--max)) * 100%),
    transparent calc((var(--value) / var(--max)) * 100%)
  );
  height: 100%;
  mask: url(./assets/dial.svg) no-repeat center;
  overflow: hidden;
  position: relative;
  width: 100%;

  &::-webkit-progress-bar {
    appearance: none;
    background: transparent;
    display: none;
  }
}
