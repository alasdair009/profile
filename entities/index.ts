export { borderRadii } from "./design-tokens/effects";
export { colors } from "./design-tokens/colors";
export { fonts } from "./design-tokens/typography";
export { sizes } from "./design-tokens/dimensions";

export { Button } from "./atoms/Button";
export { Fire } from "./atoms/Fire";
export { Heading } from "./atoms/Heading";
export { Link } from "./atoms/Link";
export { Orb } from "./atoms/Orb";
export { Paragraph } from "./atoms/Paragraph";
export { Spacer } from "./atoms/Spacer";
export { TextMask } from "./atoms/TextMask";

export { Container } from "./organisms/Container";
export { ContentPlate } from "./organisms/ContentPlate";
export { Header } from "./organisms/Header";
export { Splash } from "./organisms/Splash";
