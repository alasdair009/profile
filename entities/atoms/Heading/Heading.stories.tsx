import { Heading } from "./Heading";
import { Meta, StoryObj } from "@storybook/react";

const meta: Meta<typeof Heading> = {
  component: Heading,
  argTypes: {
    children: {
      control: {
        type: "text",
      },
    },
  },
};
export default meta;

export const Default: StoryObj<typeof Heading> = {
  args: {
    children: "Heading",
  },
};
