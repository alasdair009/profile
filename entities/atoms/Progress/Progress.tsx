import { ProgressElement, Root } from "./styles";
import { HTMLAttributes, ReactNode } from "react";

export type ProgressProps = {
  /**
   * Current progress value
   */
  value?: number;
  /**
   * Max value
   */
  max?: number;
} & HTMLAttributes<HTMLDivElement>;

/**
 * Disc to display progress to a target value
 */
export function Progress({ max = 100, value = 100, ...rest }: ProgressProps) {
  const safeValue = value > max ? max : value;
  return (
    <Root data-testid={Progress.name} {...rest}>
      <ProgressElement max={100} value={safeValue} data-testid={`${Progress.name}Element`} />
    </Root>
  );
}
