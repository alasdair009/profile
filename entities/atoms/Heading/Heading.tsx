import { Root } from "./styles";
import { HTMLAttributes, ReactNode } from "react";
import { colors, HeadingTypes, TextAlignment } from "@/entities";
import { Property } from "csstype";

export type HeadingProps = {
  /**
   * Visual style of the Heading.
   */
  level?: HeadingTypes;
  /**
   * Heading element level.
   */
  as?: HeadingTypes;
  /**
   * Contents of the heading.
   */
  children: ReactNode;
  /**
   * Text alignment of the heading.
   */
  align?: TextAlignment;
  /**
   * Color of the text.
   */
  color?: Property.Color;
  /**
   * Apply a subtle shadow behind the text.
   */
  textShadow?: boolean;
  /**
   * Number of lines the text is restricted to.
   */
  lines?: number;
} & HTMLAttributes<HTMLHeadingElement>;

/**
 * Text displayed in a heading at a customisable level visually and semantically.
 */
export function Heading({
  level = "h1",
  as,
  align = "left",
  color = colors.whiteGhost,
  textShadow = false,
  lines,
  children,
  ...rest
}: HeadingProps) {
  const headingAs = as ? as : level;
  return (
    <Root
      $color={color}
      $level={level}
      as={headingAs}
      $align={align}
      $textShadow={textShadow}
      $lines={lines}
      data-testid={Heading.name}
      {...rest}
    >
      {children}
    </Root>
  );
}
