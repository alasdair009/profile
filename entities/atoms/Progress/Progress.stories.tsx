import { Meta, StoryObj } from "@storybook/react";
import { Progress, ProgressProps } from "@/entities/atoms/Progress/Progress";
import { expect, within } from "@storybook/test";
import { Heading, Link } from "@/entities";
import { HeadingProps } from "@/entities/atoms/Heading/Heading";

const meta: Meta<typeof Progress> = {
  component: Progress,
  argTypes: {
    value: {
      control: { type: "range", min: 0, max: 100, step: 1 },
    },
    max: {
      control: { type: "range", min: 0, max: 100, step: 1 },
    },
  },
  args: {
    value: 100,
    max: 100,
  },
};
export default meta;

export const Default: StoryObj<typeof Progress> = {
  play: async ({ args, canvasElement }) => progressTest(args, canvasElement),
};
export const Zero: StoryObj<typeof Progress> = {
  args: {
    value: 0,
  },
  play: async ({ args, canvasElement }) => progressTest(args, canvasElement),
};
export const Fifty: StoryObj<typeof Progress> = {
  args: {
    value: 50,
  },
  play: async ({ args, canvasElement }) => progressTest(args, canvasElement),
};

const progressTest = async (args: ProgressProps, canvasElement: HTMLElement) => {
  const canvas = within(canvasElement);
  const progressWrapper = canvas.getByTestId(Progress.name);
  const progressElement = canvas.getByTestId(`${Progress.name}Element`);

  await expect(progressWrapper).toBeInTheDocument();
  await expect(progressElement).toBeInTheDocument();
  await expect(progressElement).toHaveAttribute("max", `${args.max}`);
  await expect(progressElement).toHaveAttribute("value", `${args.value}`);
};
