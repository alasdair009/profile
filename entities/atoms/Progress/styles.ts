import styled from "styled-components";
import { colors, sizes } from "@/entities";
import dialIcon from "./assets/dial.svg";
import { rgba } from "polished";

const getCutPoint = (value: number, max: number) => {
  return (value / max) * 100;
};

export const Root = styled.div`
  align-items: center;
  aspect-ratio: 1;
  border-radius: 50%;
  display: flex;
  height: ${sizes.s96.rem};
  justify-content: center;
  position: relative;

  &::before {
    background: radial-gradient(
      transparent,
      transparent 30%,
      ${rgba(colors.blackEvil, 0.5)} 30%,
      ${rgba(colors.blackEvil, 0.5)} 70%,
      transparent 50%
    );
    border-radius: 50%;
    content: "";
    height: 100%;
    left: 0;
    position: absolute;
    top: 0;
    width: 100%;
  }
`;
export const ProgressElement = styled.progress<{ max: number; value: number }>`
  appearance: none;
  background: conic-gradient(
    ${colors.greenGrass},
    ${colors.whiteGhost},
    ${colors.whiteGhost} ${({ max, value }) => getCutPoint(value, max)}%,
    transparent ${({ max, value }) => getCutPoint(value, max)}%
  );
  height: 100%;
  mask: url(${dialIcon.src}) no-repeat center;
  overflow: hidden;
  position: relative;
  width: 100%;

  &::-webkit-progress-bar {
    appearance: none;
    background: transparent;
    display: none;
  }
`;
