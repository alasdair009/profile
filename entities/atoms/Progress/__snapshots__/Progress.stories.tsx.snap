// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atoms/Progress Default play-test 1`] = `
<div data-testid="Progress"
     class="sc-drnuxz bSYNxs"
>
  <progress max="100"
            value="100"
            data-testid="ProgressElement"
            class="sc-hDcvty iyrkTb"
  >
  </progress>
</div>
`;

exports[`atoms/Progress Fifty play-test 1`] = `
<div data-testid="Progress"
     class="sc-drnuxz bSYNxs"
>
  <progress max="100"
            value="50"
            data-testid="ProgressElement"
            class="sc-hDcvty jkmjkn"
  >
  </progress>
</div>
`;

exports[`atoms/Progress Zero play-test 1`] = `
<div data-testid="Progress"
     class="sc-drnuxz bSYNxs"
>
  <progress max="100"
            value="0"
            data-testid="ProgressElement"
            class="sc-hDcvty cuXiGL"
  >
  </progress>
</div>
`;
